File: OldProblem/vnoi/TREECOVER.cpp
#include <bits/stdc++.h>
using namespace std;

struct FastIO {
    static const int SZ = 1 << 20;
    int in_pos = 0, in_len = 0, out_len = 0;
    char in_buf[SZ], out_buf[SZ];

    ~FastIO() { flush(); }

    char get_char() {
        if (in_pos == in_len) {
            in_len = fread(in_buf, 1, SZ, stdin);
            in_pos = 0;
            if (!in_len) return 0;
        }
        return in_buf[in_pos++];
    }

    int read_int() {
        char c = get_char();
        while (c <= ' ') c = get_char();

        int x = 0;
        while (c > ' ') {
            x = x * 10 + (c - '0');
            c = get_char();
        }
        return x;
    }

    void write_int(int x) {
        char s[12];
        int n = 0;
        do {
            s[n++] = char('0' + x % 10);
            x /= 10;
        } while (x);

        while (n--) put_char(s[n]);
        put_char('\n');
    }

    void put_char(char c) {
        if (out_len == SZ) flush();
        out_buf[out_len++] = c;
    }

    void flush() {
        if (out_len) {
            fwrite(out_buf, 1, out_len, stdout);
            out_len = 0;
        }
    }
};

struct Fenwick {
    int n, mask = 1;
    vector<int> bit;

    Fenwick(int n = 0) : n(n), bit(n + 1) {
        while ((mask << 1) <= n) mask <<= 1;
    }

    void add(int i, int v) {
        for (; i <= n; i += i & -i) bit[i] += v;
    }

    int sum(int i) const {
        int s = 0;
        for (; i; i -= i & -i) s += bit[i];
        return s;
    }

    int kth(int k) const {
        int i = 0;
        for (int step = mask; step; step >>= 1) {
            int nxt = i + step;
            if (nxt <= n && bit[nxt] < k) {
                i = nxt;
                k -= bit[nxt];
            }
        }
        return i + 1;
    }
};

struct ActiveSet {
    int n, blocks, non_empty = 0;
    vector<unsigned long long> bits;
    Fenwick fw;

    ActiveSet(int n = 0) : n(n), blocks((n + 63) >> 6), bits(blocks), fw(blocks) {}

    void add(int pos) {
        int b = (pos - 1) >> 6, off = (pos - 1) & 63;
        if (!bits[b]) {
            fw.add(b + 1, 1);
            ++non_empty;
        }
        bits[b] |= 1ULL << off;
    }

    void remove(int pos) {
        int b = (pos - 1) >> 6, off = (pos - 1) & 63;
        bits[b] &= ~(1ULL << off);
        if (!bits[b]) {
            fw.add(b + 1, -1);
            --non_empty;
        }
    }

    int first_pos() const {
        int b = fw.kth(1) - 1;
        return (b << 6) + __builtin_ctzll(bits[b]) + 1;
    }

    int last_pos() const {
        int b = fw.kth(non_empty) - 1;
        return (b << 6) + 64 - __builtin_clzll(bits[b]);
    }

    int next_pos(int pos) const {
        int idx = pos, b = idx >> 6, off = idx & 63;
        if (b >= blocks) return 0;

        if (b < blocks) {
            unsigned long long cur = bits[b] & (~0ULL << off);
            if (cur) return (b << 6) + __builtin_ctzll(cur) + 1;
        }

        int before = fw.sum(b + 1);
        if (before == non_empty) return 0;
        b = fw.kth(before + 1) - 1;
        return (b << 6) + __builtin_ctzll(bits[b]) + 1;
    }

    int prev_pos(int pos) const {
        int idx = pos - 2;
        if (idx < 0) return 0;

        int b = idx >> 6, off = idx & 63;
        unsigned long long mask = off == 63 ? ~0ULL : ((1ULL << (off + 1)) - 1);
        unsigned long long cur = bits[b] & mask;
        if (cur) return (b << 6) + 64 - __builtin_clzll(cur);

        int before = fw.sum(b);
        if (!before) return 0;
        b = fw.kth(before) - 1;
        return (b << 6) + 64 - __builtin_clzll(bits[b]);
    }
};

int main() {
    static FastIO io;
    int n = io.read_int();
    int t = io.read_int();

    vector<int> head(n + 1, -1), to(2 * max(0, n - 1)), nxt(2 * max(0, n - 1));
    int edges = 0;

    auto add_edge = [&](int u, int v) {
        to[edges] = v;
        nxt[edges] = head[u];
        head[u] = edges++;
    };

    for (int i = 1; i < n; ++i) {
        int u = io.read_int();
        int v = io.read_int();
        add_edge(u, v);
        add_edge(v, u);
    }

    vector<int> parent(n + 1), depth(n + 1), tin(n + 1), tout(n + 1), order(n + 1);
    vector<int> st;
    st.reserve(n);
    st.push_back(1);

    int timer = 0;
    while (!st.empty()) {
        int u = st.back();
        st.pop_back();

        tin[u] = tout[u] = ++timer;
        order[timer] = u;

        for (int e = head[u]; e != -1; e = nxt[e]) {
            int v = to[e];
            if (v == parent[u]) continue;
            parent[v] = u;
            depth[v] = depth[u] + 1;
            st.push_back(v);
        }
    }

    for (int i = n; i > 1; --i) {
        int u = order[i];
        int p = parent[u];
        if (tout[u] > tout[p]) tout[p] = tout[u];
    }

    int lg = 32 - __builtin_clz(n);
    int stride = n + 1;
    vector<int> up(lg * stride);

    for (int u = 1; u <= n; ++u) up[u] = parent[u];
    for (int i = 1; i < lg; ++i) {
        int base = i * stride;
        int prev = base - stride;
        for (int u = 1; u <= n; ++u) up[base + u] = up[prev + up[prev + u]];
    }

    auto ancestor = [&](int u, int v) {
        return tin[u] <= tin[v] && tin[v] <= tout[u];
    };

    auto child_below = [&](int u, int x) {
        int diff = depth[x] - depth[u] - 1;
        for (int i = 0; diff; ++i, diff >>= 1) {
            if (diff & 1) x = up[i * stride + x];
        }
        return x;
    };

    ActiveSet active_pos(n);
    vector<unsigned char> active(n + 1), leaf(n + 1);
    int active_count = 0, leaf_count = 0;

    auto prev_node = [&](int pos) {
        int p = active_pos.prev_pos(pos);
        return order[p ? p : active_pos.last_pos()];
    };

    auto next_node = [&](int pos) {
        int p = active_pos.next_pos(pos);
        return order[p ? p : active_pos.first_pos()];
    };

    auto check_leaf = [&](int u, int pre, int nxt_node) {
        if (active_count <= 2) return true;

        bool in_pre = ancestor(u, pre);
        bool in_next = ancestor(u, nxt_node);
        if (!in_pre && !in_next) return true;
        if (in_pre != in_next) return false;
        return child_below(u, pre) == child_below(u, nxt_node);
    };

    auto set_leaf = [&](int u, bool now) {
        if (now != leaf[u]) {
            leaf_count += now ? 1 : -1;
            leaf[u] = (unsigned char)now;
        }
    };

    while (t--) {
        int u = io.read_int();
        int pos = tin[u];

        if (active[u]) {
            int total = active_count;
            int pre = 0, nxt_node = 0, pre2 = 0, nxt2 = 0;

            if (total > 1) {
                pre = prev_node(pos);
                nxt_node = next_node(pos);
                if (total > 3) {
                    pre2 = prev_node(tin[pre]);
                    nxt2 = next_node(tin[nxt_node]);
                }
            }

            leaf_count -= leaf[u];
            leaf[u] = active[u] = 0;
            --active_count;
            active_pos.remove(pos);

            if (active_count == 1) {
                set_leaf(pre, true);
            } else if (active_count == 2) {
                set_leaf(pre, true);
                set_leaf(nxt_node, true);
            } else if (active_count > 2) {
                set_leaf(pre, check_leaf(pre, pre2, nxt_node));
                set_leaf(nxt_node, check_leaf(nxt_node, pre, nxt2));
            }
        } else {
            active[u] = 1;
            ++active_count;
            active_pos.add(pos);

            if (active_count == 1) {
                set_leaf(u, true);
            } else if (active_count == 2) {
                int other = prev_node(pos);
                set_leaf(u, true);
                set_leaf(other, true);
            } else {
                int pre = prev_node(pos);
                int nxt_node = next_node(pos);
                int pre2 = prev_node(tin[pre]);
                int nxt2 = next_node(tin[nxt_node]);

                set_leaf(u, check_leaf(u, pre, nxt_node));
                set_leaf(pre, check_leaf(pre, pre2, u));
                set_leaf(nxt_node, check_leaf(nxt_node, u, nxt2));
            }
        }

        io.write_int((leaf_count + 1) >> 1);
    }

    return 0;
}

File: OldProblem/vnoi/TREECOVER.py
import sys


class Fenwick:
    def __init__(self, n: int):
        self.n = n
        self.bit = [0] * (n + 1)
        self.mask = 1
        while (self.mask << 1) <= n:
            self.mask <<= 1

    def add(self, idx: int, delta: int) -> None:
        while idx <= self.n:
            self.bit[idx] += delta
            idx += idx & -idx

    def sum(self, idx: int) -> int:
        total = 0
        while idx:
            total += self.bit[idx]
            idx -= idx & -idx
        return total

    def kth(self, k: int) -> int:
        idx = 0
        step = self.mask
        while step:
            nxt = idx + step
            if nxt <= self.n and self.bit[nxt] < k:
                idx = nxt
                k -= self.bit[nxt]
            step >>= 1
        return idx + 1


class ActiveSet:
    """Store active DFS positions by 64-position blocks."""

    def __init__(self, n: int):
        self.n = n
        self.blocks = (n + 63) >> 6
        self.bits = [0] * self.blocks
        self.non_empty = 0
        self.fw = Fenwick(self.blocks)

    @staticmethod
    def first_bit(mask: int) -> int:
        return (mask & -mask).bit_length() - 1

    def add(self, pos: int) -> None:
        idx = pos - 1
        block = idx >> 6
        bit = 1 << (idx & 63)

        if self.bits[block] == 0:
            self.fw.add(block + 1, 1)
            self.non_empty += 1
        self.bits[block] |= bit

    def remove(self, pos: int) -> None:
        idx = pos - 1
        block = idx >> 6
        bit = 1 << (idx & 63)

        self.bits[block] &= ~bit
        if self.bits[block] == 0:
            self.fw.add(block + 1, -1)
            self.non_empty -= 1

    def first_pos(self) -> int:
        block = self.fw.kth(1) - 1
        return (block << 6) + self.first_bit(self.bits[block]) + 1

    def last_pos(self) -> int:
        block = self.fw.kth(self.non_empty) - 1
        return (block << 6) + self.bits[block].bit_length()

    def next_pos(self, pos: int) -> int:
        # Strictly after pos.
        idx = pos
        block = idx >> 6
        offset = idx & 63
        if block >= self.blocks:
            return 0

        cur = self.bits[block] & (-1 << offset)
        if cur:
            return (block << 6) + self.first_bit(cur) + 1

        before = self.fw.sum(block + 1)
        if before == self.non_empty:
            return 0

        block = self.fw.kth(before + 1) - 1
        return (block << 6) + self.first_bit(self.bits[block]) + 1

    def prev_pos(self, pos: int) -> int:
        # Strictly before pos.
        idx = pos - 2
        if idx < 0:
            return 0

        block = idx >> 6
        offset = idx & 63
        cur = self.bits[block] & ((1 << (offset + 1)) - 1)
        if cur:
            return (block << 6) + cur.bit_length()

        before = self.fw.sum(block)
        if before == 0:
            return 0

        block = self.fw.kth(before) - 1
        return (block << 6) + self.bits[block].bit_length()


def solve() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return

    ptr = 0
    n = data[ptr]
    q = data[ptr + 1]
    ptr += 2

    adj = [[] for _ in range(n + 1)]
    for _ in range(n - 1):
        u = data[ptr]
        v = data[ptr + 1]
        ptr += 2
        adj[u].append(v)
        adj[v].append(u)

    parent = [0] * (n + 1)
    depth = [0] * (n + 1)
    tin = [0] * (n + 1)
    tout = [0] * (n + 1)
    order = [0] * (n + 1)

    stack = [1]
    timer = 0
    while stack:
        u = stack.pop()
        timer += 1
        tin[u] = tout[u] = timer
        order[timer] = u

        for v in adj[u]:
            if v == parent[u]:
                continue
            parent[v] = u
            depth[v] = depth[u] + 1
            stack.append(v)

    for idx in range(n, 1, -1):
        u = order[idx]
        p = parent[u]
        if tout[u] > tout[p]:
            tout[p] = tout[u]

    log = n.bit_length()
    up = [parent[:]]
    for level in range(1, log):
        prev = up[level - 1]
        cur = [0] * (n + 1)
        for u in range(1, n + 1):
            cur[u] = prev[prev[u]]
        up.append(cur)

    def ancestor(u: int, v: int) -> bool:
        return tin[u] <= tin[v] <= tout[u]

    def child_below(u: int, x: int) -> int:
        # x is inside u subtree; return the direct child of u that contains x.
        diff = depth[x] - depth[u] - 1
        bit = 0
        while diff:
            if diff & 1:
                x = up[bit][x]
            diff >>= 1
            bit += 1
        return x

    active_pos = ActiveSet(n)
    active = bytearray(n + 1)
    leaf = bytearray(n + 1)
    active_count = 0
    leaf_count = 0

    def prev_node(pos: int) -> int:
        prev_pos = active_pos.prev_pos(pos)
        return order[prev_pos if prev_pos else active_pos.last_pos()]

    def next_node(pos: int) -> int:
        nxt_pos = active_pos.next_pos(pos)
        return order[nxt_pos if nxt_pos else active_pos.first_pos()]

    def check_leaf(u: int, pre: int, nxt: int) -> bool:
        if active_count <= 2:
            return True

        in_pre = ancestor(u, pre)
        in_next = ancestor(u, nxt)
        if not in_pre and not in_next:
            return True
        if in_pre != in_next:
            return False
        return child_below(u, pre) == child_below(u, nxt)

    def set_leaf(u: int, now: bool) -> None:
        nonlocal leaf_count
        if now != bool(leaf[u]):
            leaf_count += 1 if now else -1
            leaf[u] = 1 if now else 0

    out = []
    for _ in range(q):
        u = data[ptr]
        ptr += 1
        pos = tin[u]

        if active[u]:
            total = active_count
            pre = nxt = pre2 = nxt2 = 0

            if total > 1:
                pre = prev_node(pos)
                nxt = next_node(pos)
                if total > 3:
                    pre2 = prev_node(tin[pre])
                    nxt2 = next_node(tin[nxt])

            leaf_count -= leaf[u]
            leaf[u] = 0
            active[u] = 0
            active_count -= 1
            active_pos.remove(pos)

            if active_count == 1:
                set_leaf(pre, True)
            elif active_count == 2:
                set_leaf(pre, True)
                set_leaf(nxt, True)
            elif active_count > 2:
                set_leaf(pre, check_leaf(pre, pre2, nxt))
                set_leaf(nxt, check_leaf(nxt, pre, nxt2))
        else:
            active[u] = 1
            active_count += 1
            active_pos.add(pos)

            if active_count == 1:
                set_leaf(u, True)
            elif active_count == 2:
                other = prev_node(pos)
                set_leaf(u, True)
                set_leaf(other, True)
            else:
                pre = prev_node(pos)
                nxt = next_node(pos)
                pre2 = prev_node(tin[pre])
                nxt2 = next_node(tin[nxt])

                set_leaf(u, check_leaf(u, pre, nxt))
                set_leaf(pre, check_leaf(pre, pre2, u))
                set_leaf(nxt, check_leaf(nxt, u, nxt2))

        # A tree with L leaves can be covered by ceil(L / 2) paths.
        out.append(str((leaf_count + 1) >> 1))

    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    solve()
